Recover from corrupt cache metadata or config file

wjt opened the pull request:
We observed a system where `/etc/metrics/cache-size.conf` and `/var/cache/metrics/variants.dat.metadata` were full of `NUL` bytes, to the length you'd expect if the file were properly written. When it encountered these files, the metrics daemon logged a warning for the first, then logged a fatal error and fell over for the second.

I believe this is due to surprising behaviour of `g_file_set_contents()` when the file to be overwritten does not exist: it does not `fsync()` the temporary file before `close()` and `rename()`. Is this a bug? Maybe. This application appears to be doing everything right (relatively speaking): it creates both files with `g_key_file_save_to_file()` which is meant to be the correct, safe thing to do.

Regardless of the cause, any systems in the wild which have hit this problem now have a broken metrics daemon, so we need to deal with the case where one or both of these keyfiles cannot be parsed. The sensible recovery options seem to me to be:

* `cache-size.conf`: since there's only one value in this file, just re-initialize it.
* `variants.dat.metadata`: without the metadata it's impossible to interpret the contents of `variants.dat`, so again we need to re-initialize the keyfile, which logically discards the contents of `variants.dat`. In this case, we want to log an event so we have some idea how often this happens in the wild.

`EmerCacheSizeProvider` was, to my eyes, really overcomplicated, so I'm afraid I couldn't stop myself replacing it with a single function.

https://phabricator.endlessm.com/T19953

wjt commented on daemon/emer-cache-size-provider.c:
```diff
@@ -44,199 +34,55 @@ G_DEFINE_TYPE_WITH_PRIVATE (EmerCacheSizeProvider, emer_cache_size_provider, G_T
-        goto handle_failed_read;
-
+        {
+          g_warning ("Error reading cache size from %s: %s", path,
```
Looking again, I think we should ignore `G_KEY_FILE_ERROR_GROUP_NOT_FOUND` and `G_KEY_FILE_ERROR_KEY_NOT_FOUND` and just silently add the key.

wjt commented on daemon/emer-circular-file.c:
```diff
@@ -490,19 +495,23 @@ emer_circular_file_initable_init (GInitable    *initable,
+  if (!priv->reinitialize &&
+      !g_key_file_load_from_file (priv->metadata_key_file,
+                                  priv->metadata_filepath, G_KEY_FILE_NONE,
+                                  &local_error))
```
Similarly here?

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -103,8 +103,10 @@ make_circular_file (Fixture *fixture,
                     guint64  max_size)
 {
   GError *error = NULL;
+  gboolean reinitialize = FALSE;
```
It looks like you can't avoid this declaration and pass `FALSE` directly to `_new() `

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
       const gchar * const argv[])
 {
   g_test_init (&argc, (gchar ***) &argv, NULL);
+  g_test_bug_base ("https://phabricator.endlessm.com/");
```
nice

wjt replied:
Yeah, it is nice. If you call `g_test_bug("T19953")` in a test and run the test suite with `--verbose` it shows up in the output:

```
GTest: run: /cache-size-provider/recovers-if-corrupt/nul
(MSG: Bug Reference: https://phabricator.endlessm.com/T19953)
GTest: result: OK
```

Whether it's actually more useful than a comment is another question but it's there so let's use it!

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+  g_autoptr(EmerCircularFile) circular_file = NULL;
+
+  circular_file = emer_circular_file_new (fixture->data_file_path, max_size,
+                                          /* reinitialize */ FALSE,
```
Nit: I normally see this inline comments added **to the right** of the value, not to the left, because of keeping everything aligned on the left, so maybe do it here too?. Mind you, it might be that I'm used to a different style, so feel free to ignore

wjt replied:
Fair enough: the style you prefer is also the style used elsewhere in this file so I will adjust my style.

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+  g_clear_error (&error);
+
+  circular_file = emer_circular_file_new (fixture->data_file_path, max_size,
+                                          /* reinitialize */ TRUE,
```
Same comment (nit) here

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+  const gchar * const STRINGS[] =
+    {
+      "Sneezy", "phylum", "Europe", "sloth", "guacamole", "data link",
+      "Colossus of Rhodes"
```
May I suggest "Mint cake" too?

wjt replied:
A very good idea

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+   * of asterisks.
+   */
+  const gchar * const NO_STRINGS[0];
+  guint64 token = assert_strings_read (circular_file, NO_STRINGS, 0);
```
I know this is not your code, and also it could be due to me being used to other naming conventions, but I found it surprisingly confusing that `assert_strings_read()` does actually more stuff than simply asserting the desired outcome/state. It would be more clear, IMHO, if it was named something like `read_strings_and_check()` or similar, so that it's clear that it performs both the action that needs to be tested + the actual test.

Not for this patch though, but maybe to consider for a separate cleanup commit?

wjt replied:
Yes, great idea.

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+      "Colossus of Rhodes"
+    };
+  gsize NUM_STRINGS = G_N_ELEMENTS (STRINGS);
+  assert_strings_saved (circular_file, STRINGS, NUM_STRINGS);
```
Same comment than for `assert_strings_read()` here (consider renaming in a separate cleanup commit?)

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+   * consider the circular file itself to be empty.
+   */
+  test_circular_file_broken_metadata_file (fixture, "", 0,
+                                           G_KEY_FILE_ERROR_GROUP_NOT_FOUND);
```
You'll have to update the expectation here if you finally silently deal with empty files and add the missing group/keys there, I think.

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+  g_test_bug ("T19953");
+
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_GROUP_NOT_FOUND);
```
Same here

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+   * edits the file by hand and then you get to keep both pieces.
+   */
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_KEY_NOT_FOUND);
```
And here too

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+   * we do not know at what byte we should loop around.
+   */
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_KEY_NOT_FOUND);
```
Likewise

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+  circular_file = emer_circular_file_new (fixture->data_file_path, max_size,
+                                          /* reinitialize */ FALSE,
+                                          &error);
+  g_assert_null (circular_file);
```
If you silently deal with missing groups/keys you'll receive a valid circular file here, so these assertions won't longer be valid, I think. Might need to refactor this code a bit in that case

wjt replied:
Good point – I'll split this test helper in two.

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+   * we don't know where the head was supposed to be.
+   */
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_KEY_NOT_FOUND);
```
And here as well. Actually, it looks like you might need to refactor these tests a bit to accomodate the different way of dealing with missing groups/keys, because the `expected_error` parameter seems like not very useful except for the "junk data" case :-)

wjt replied:
Right – I decided that this case and the ones below, when the file's actively malformed (rather than being empty or all `NUL` which is "morally empty"), are serious enough that we do still want to return an error, and have the daemon report an event and explicitly reinitialize. So I split this up.

mariospr replied:
Sounds good to me

mariospr commented on daemon/emer-cache-size-provider.c:
```diff
@@ -44,199 +34,61 @@ G_DEFINE_TYPE_WITH_PRIVATE (EmerCacheSizeProvider, emer_cache_size_provider, G_T
+  if (!g_key_file_load_from_file (key_file, path, G_KEY_FILE_NONE,
+                                  &error))
     {
       if (!g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
```
Following a similar pattern in other places (based on your own review), you might want to also exclude here cases where the file exists but the either the GROUP or the KEY can't be found, in which case you'd add default values?

wjt replied:
Yes, I'll make that change.

wjt commented:
I've made the necessary changes to handle the "empty keyfile" and "all-NUL keyfile" cases silently, while still warning/returning an error in the "actively malformed" cases. I also think I've addressed your feedback – thanks!

mariospr commented on tests/daemon/test-circular-file.c:
```diff
@@ -595,25 +597,249 @@ test_circular_file_shrink (Fixture      *fixture,
+  g_assert_false (emer_circular_file_has_more (circular_file, token));
+
+  /* Adding new entries to the file and reading them back should work. */
+  const gchar * const STRINGS[] = { "Kendal Mint Cake" };
```
EXCELLENT! /cc @pwithnall

mariospr commented:
Thanks @wjt, I looked through the differences (with special emphasis on the updated test cases) and it all looks good to me now. Merging...
